Added chained comparison

Dherse opened the pull request:
Closes #2184.

If this feature is accepted (emphasis on the **if**), this PR would add the range comparison operator similar to Python: `1 < a <= 2`.

It it purposefully limited to the "basic" comparison operator `<, >, <=, >=, ==, !=`.

This simple changes passes all tests, and new ones are added to test this new functionality.

I mostly implemented it to get my feet wet with the AST and parsing since I haven't had a look before 👏

MDLC01 commented:
Although I really like chained comparisons, I think it should only be allowed in cases where transitivity is preserved, meaning a chain can only contain:
- `<`, `<=` and `==`, or
- `>`, `>=` and `==`.

In other words, a chain only makes sense, in my opinion, when you can infer a relationship between the first and the last element of the chain. For example, `a < b == c <= d` is fine, but `a > b < c` is not.

This ensures it is not possible to write something like `a < b > c == d != e`, which is impossible to understand, or, worse, `a != b != c`, which seems to imply `a != c`, although it does not.

Dherse commented:
I fully agree with you, I think this makes the most sense and prevents misuse. I'll try implementing it in the morning, it should be fairly easy

Dherse commented:
@MDLC01 I added a transitiveness check (please see below), and I also added a check that you cannot add a third (or more) operations after which would cause a runtime error either way (cannot compare bool and int for example)

The code for the erroneous states (less numerous than correct ones) is the following, could you check it please:

```rs
match (op, second_op) {
    (ast::BinOp::Gt, ast::BinOp::Lt)
    | (ast::BinOp::Gt, ast::BinOp::Leq)
    | (ast::BinOp::Geq, ast::BinOp::Lt)
    | (ast::BinOp::Geq, ast::BinOp::Leq)
    | (ast::BinOp::Lt, ast::BinOp::Gt)
    | (ast::BinOp::Lt, ast::BinOp::Geq)
    | (ast::BinOp::Leq, ast::BinOp::Gt)
    | (ast::BinOp::Leq, ast::BinOp::Geq)
    | (ast::BinOp::Neq, ast::BinOp::Neq) => {
        if let Some(node) = p.node_mut(m) {
            node.convert_to_error(
                "only transitive comparisons are allowed",
            );
        }
    }
    _ => {
        p.wrap(m, SyntaxKind::TernaryComp);
        continue;
    }
}
```

laurmaedje commented:
I think a better terminology would be "chained comparison" across the board. Ternary sounds like `a ? b : c`.

laurmaedje commented on tests/typ/compiler/ops.typ:
```diff
@@ -301,3 +301,41 @@
+---
+// Test invalid chained operations
+
+// Error: 7-8 only transitive comparisons are allowed
+#test(1 > 2 < 3)
+
+#let (a, b, c, d, e) = (1, 2, 3, 4, 5)
+
+// Error: 7-8 can only chain at most two comparisons
+#test(a < b > c == d != e)
```
the error spans seem not ideal. they should cover the whole comparison or at least the faulty operator, not the first operand.

laurmaedje commented on crates/typst/src/eval/mod.rs:
```diff
@@ -1076,23 +1077,71 @@ impl Eval for ast::Binary<'_> {
+}
+
+/// Apply a basic binary operation.
+fn apply_binary_op<'a, L, R>(
```
why all the closures here? why not just `&ast::Expr`?

Dherse replied:
That's what I changed it too. I was under the (wrong) assumption that the reason why the `Binary` was passed to the closure was somehow for performance reason and I wanted to keep the getting of `lhs` or `rhs` as late as possible, but it doesn't make much sense in hindsight. That's what I get for opening a PR at like midnight 😅

laurmaedje commented on crates/typst/src/eval/mod.rs:
```diff
@@ -1076,23 +1077,71 @@ impl Eval for ast::Binary<'_> {
+            _ => bail!(
+                self.span(),
+                "{} is not a valid comparison operator",
+                self.lhs_op().as_str()
```
why lhs op?

Dherse replied:
A small mistake, fixed

laurmaedje commented on crates/typst/src/eval/mod.rs:
```diff
@@ -1076,23 +1077,71 @@ impl Eval for ast::Binary<'_> {
+
+    #[tracing::instrument(name = "ChainedComp::eval", skip_all)]
+    fn eval(self, vm: &mut Vm) -> SourceResult<Self::Output> {
+        let mut side = |lhs: &dyn Fn() -> ast::Expr<'a>,
+                        op,
+                        rhs: &dyn Fn() -> ast::Expr<'a>| match op {
```
I think this is a case where dyn Fn could actually be worse than generics. besides the match could just map to `ops::..` instead of the whole apply thing. in general, I feel like the evaluation part could be simplified a bit.

laurmaedje commented on crates/typst-syntax/src/ast.rs:
```diff
@@ -1372,6 +1376,46 @@ impl<'a> Binary<'a> {
 
+node! {
+    /// A chained comparison operation: `a < b <= c`.
+    ChainedComp
```
```suggestion
    ChainedComparison
```

laurmaedje commented on crates/typst-syntax/src/ast.rs:
```diff
@@ -1515,6 +1573,11 @@ impl BinOp {
     }
+
+    /// Whether this operator is a comparison operator.
+    pub fn is_comp(self) -> bool {
```
```suggestion
    pub fn is_comparison(self) -> bool {
```

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -687,6 +687,60 @@ fn code_expr_prec(
+                            }
+                        }
+                        _ => {
+                            p.wrap(m, SyntaxKind::ChainedComp);
+                            continue;
```
I think the wrapping should be the last thing that happens. This makes for clearer control flow.

Dherse replied:
Ok, I think it should be fixed, please let me know. (once I have pushed the commit)

Dherse commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -687,6 +687,60 @@ fn code_expr_prec(
             code_expr_prec(p, false, prec, false);
+
+            // Special case for chained comparison operations.
+            if op.is_comp() && ast::BinOp::NotIn.precedence() >= min_prec {
```
Me trying to understand what was going on while tired, it's fixed now ;)

istudyatuni commented on tests/typ/compiler/ops.typ:
```diff
@@ -301,3 +301,41 @@
+// Test invalid chained operations
+
+// Error: 14-16 only transitive comparisons are allowed
+#test(1 != 2 != 3, true)
```
I can't understand, why this is not allowed? Or, maybe there exists a more clear example:

```suggestion
#test(1 != 2 != 1, false)
```

```
1 != 2 -> true
2 != 1 -> true
but 1 != 1 is false
```

If I understand transitivity correctly

Dherse replied:
The boolean value is kind of irrelevant as this is not allowed as per @MDLC01's comment

istudyatuni commented on tests/typ/compiler/ops.typ:
```diff
@@ -301,3 +301,41 @@
 #test(inc2(2, y: 4), 6)
+
+---
+// Test chained comparison operations
```
Could you also add

```typ
#test((1 < 2) == (2 < 1), false)
```

I think it should work

Dherse replied:
But that's not really in the scope of this PR tho, since it is valid syntax already before hand

laurmaedje commented:
Concerns were raised about the interaction between this and operator overloading. Operator overloading isn't currently a thing (and I'm not yet convinced that it should in the future), but we should keep the possibility in mind here.

PgBiel commented:
> Concerns were raised about the interaction between this and operator overloading. Operator overloading isn't currently a thing (and I'm not yet convinced that it should in the future), but we should keep the possibility in mind here.

If this is ever added, then it should be the responsibility of the operator overloading user to ensure the comparison operations are consistent, and `a < b` and `a > b` can never both be true.

MDLC01 commented:
Which can be done by only letting them overload `=` and, say, `<`. That way, the compiler can derive itself the implementations of `<=`, `>` and `>=`.

Jollywatt commented:
A minor point, but I’d put a vote in for allowing `a > b < c` and similar chained comparisons, like Python and Julia, for example.

The interpretation is fairly unambiguous, and it would be less complex and opinionated to simply allow it. It would also avoid the need for a transitivity check, which would require hard-coding assumptions about the runtime behaviour of `a < b` and `a > b` into the syntax, as pointed out earlier.

Enivex commented:
> A minor point, but I’d put a vote in for allowing `a > b < c` and similar chained comparisons. Python and Julia allow this, for example

I vote no

The only reason to allow this would be overloading < and > to something that doesn't act like one expects < and > to act, and that opens an entire can of worms.

laurmaedje commented:
I think I am okay with accepting this. The main concern is operator overloading, but overloading a comparison operator to return something other than a bool is strange and would likely not be supported anyway. One thing I'd like to see before merging though is some change to make `code_expr_prec` in the parser less huge. There's got to be some way to extract a function there. If it turns out that this makes the code worse (because of continue usage for instance), so be it, but we should try.

Dherse commented:
While this is pretty much merge-ready, as discussed on Discord, this introduces extra bloat, extra syntax, and more risk of breaking stuff in the future with operator overloading. Time to sunset this idea 🌞
